evalml/pipelines/components/transformers/encoders/ordinal_encoder.py

```diff
@@ -0,0 +1,320 @@
+            )
+        if handle_unknown == "use_encoded_value" and unknown_value is None:
+            raise ValueError(
+                "To use encoded value for unknown categories, unknown_value must"
+                "be specified as either np.nan or as an int that is distinct from"
+                "the other encoded categories ",
+            )
+        if top_n is not None and categories is not None:
+            raise ValueError("Cannot use categories and top_n arguments simultaneously")
+
+        self.features_to_encode = features_to_encode
+        self._encoder = None
+
+        super().__init__(
+            parameters=parameters,
+            component_obj=None,
+            random_seed=random_seed,
+        )
+        self._initial_state = self.random_seed
+        self._provenance = {}
+
+    @staticmethod
+    def _get_ordinal_cols(X):
+        """Get names of ordinal columns in the input DataFrame."""
+        return list(X.ww.select(include=["ordinal"], return_schema=True).columns)
+
+    def fit(self, X, y=None):
+        """Fits the ordinal encoder component.
+
+        Args:
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features].
+            y (pd.Series, optional): The target training data of length [n_samples].
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: If encoding a column failed.
+            TypeError: If non-Ordinal columns are specified in features_to_encode.
+        """
+        top_n = self.parameters["top_n"]
+        # Ordinal type is not inferred by Woodwork, so if it wasn't set before, it won't be set at init
+        X = infer_feature_types(X)
+        if self.features_to_encode is None:
+            self.features_to_encode = self._get_ordinal_cols(X)
+        else:
+            # When features_to_encode is user-specified, check that all columns are present
+            # and have the Ordinal logical type
+            not_present_features = [
+                col for col in self.features_to_encode if col not in list(X.columns)
+            ]
+            if len(not_present_features) > 0:
+                raise ValueError(
+                    "Could not find and encode {} in input data.".format(
+                        ", ".join(not_present_features),
+                    ),
+                )
+
+            logical_types = X.ww.logical_types
+            for col in self.features_to_encode:
+                ltype = logical_types[col]
+                if not isinstance(ltype, Ordinal):
+                    raise TypeError(
+                        f"Column {col} specified in features_to_encode is not Ordinal in nature",
+                    )
+
+            # Put features_to_encode in the same relative order as the columns in the dataframe
+            self.features_to_encode = [
+                col for col in X.columns if col in self.features_to_encode
+            ]
```

Comment:
This is in place because we say that 
> `categories[i]` is a list of the categories for the column at index `i` in the dataframes passed in at fit and transform

That kind of breaks down once you're specifying `features_to_encode` or if not all of the dataframe's columns are ordinal in nature, so I'm making the assumption that the order of the sublists in `categories` is just the relative order from the original dataframe. 

I can make note of that assumption in the docstring. Alternatively, we could just make `categories` a dictionary mapping column name to the list of categories--was there a specific reason it's a list of lists other than that's what we pass into the SKLearn encoder? The SKLearn encoder requires that categories be the same length as the number of columns in the dataframe, and now that I really think about it, a dictionary probably fits our use case better.

Reply:
I'll defer to you on this one! If a dictionary makes more sense I support switching over, but this is just fine as is.

Reply:
Okay, I went through the refactor to use a dict, and I continue to be a fan. It wasn't a small number of changes, though:  54386f0fa

It lets users avoid having to worry about the relative order of their inputs to the encoder and lets us remove this line.